jetty-server/src/main/java/org/eclipse/jetty/server/AsyncAttributes.java

```diff
@@ -0,0 +1,130 @@
+//
+//  ========================================================================
+//  Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.server;
+
+import java.util.HashSet;
+import java.util.Set;
+import javax.servlet.AsyncContext;
+
+import org.eclipse.jetty.util.Attributes;
+
+class AsyncAttributes extends Attributes.Wrapper
+{
+    /**
+     * Async dispatch attribute name prefix.
+     */
+    public static final String __ASYNC_PREFIX = "javax.servlet.async.";
+
+    private String _requestURI;
+    private String _contextPath;
+    private String _servletPath;
+    private String _pathInfo;
+    private String _query;
+
+    AsyncAttributes(Attributes attributes)
+    {
+        super(attributes);
+    }
+
+    @Override
+    public Object getAttribute(String key)
+    {
+        if (!key.startsWith(__ASYNC_PREFIX))
+            return super.getAttribute(key);
+
+        switch (key)
+        {
+            case AsyncContext.ASYNC_REQUEST_URI:
+                return _requestURI;
+            case AsyncContext.ASYNC_CONTEXT_PATH:
+                return _contextPath;
+            case AsyncContext.ASYNC_SERVLET_PATH:
+                return _servletPath;
+            case AsyncContext.ASYNC_PATH_INFO:
+                return _pathInfo;
+            case AsyncContext.ASYNC_QUERY_STRING:
+                return _query;
+            default:
+                return super.getAttribute(key);
+        }
+    }
+
+    @Override
+    public Set<String> getAttributeNameSet()
+    {
+        HashSet<String> set = new HashSet<>();
+        for (String name : _attributes.getAttributeNameSet())
+        {
+            if (!name.startsWith(__ASYNC_PREFIX))
+                set.add(name);
+        }
+
+        if (_requestURI != null)
+            set.add(AsyncContext.ASYNC_REQUEST_URI);
+        if (_contextPath != null)
+            set.add(AsyncContext.ASYNC_CONTEXT_PATH);
+        if (_servletPath != null)
+            set.add(AsyncContext.ASYNC_SERVLET_PATH);
+        if (_pathInfo != null)
+            set.add(AsyncContext.ASYNC_PATH_INFO);
+        if (_query != null)
+            set.add(AsyncContext.ASYNC_QUERY_STRING);
+
+        return set;
+    }
+
+    @Override
+    public void setAttribute(String key, Object value)
```

Comment:
Currently when we create `AsyncContextEvent` it does:
```java
                baseRequest.setAttribute(AsyncContext.ASYNC_REQUEST_URI, baseRequest.getRequestURI());
                baseRequest.setAttribute(AsyncContext.ASYNC_CONTEXT_PATH, baseRequest.getContextPath());
                baseRequest.setAttribute(AsyncContext.ASYNC_SERVLET_PATH, baseRequest.getServletPath());
                baseRequest.setAttribute(AsyncContext.ASYNC_PATH_INFO, baseRequest.getPathInfo());
                baseRequest.setAttribute(AsyncContext.ASYNC_QUERY_STRING, baseRequest.getQueryString());
```
So that's 5 (6 in jetty-10) `startWith` and string switches.    It would be good to be able set these in bulk.
So either the `AsyncContextEvent` can unwrap and downcast until it gets the `AsyncAttributesMap` and directly sets these fields... or have a synthentic set-only attribute called "org.eclipse.jetty.servlet.AsyncContextEvent.request" and pass it the request object.  This class then can extract all 5 (eventually 6) fields directly from the request. Ditto for the forwarded path with another synth attribute.